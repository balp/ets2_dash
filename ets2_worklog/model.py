import datetime
import json
from dataclasses import dataclass
from typing import Optional, Dict, List

import ets2.model


@dataclass
class Delivered:
    auto_load: bool
    auto_park: bool
    cargo_damage: float
    time: int
    distance: float
    xp: int
    revenue: int


def delivered_from_dict(data: Dict) -> Delivered:
    return Delivered(auto_load=data['auto.load.used'], auto_park=data['auto.park.used'],
                     cargo_damage=data['cargo.damage'], time=data['delivery.time'],
                     distance=data['distance.km'], xp=data['earned.xp'], revenue=data['revenue'])


@dataclass
class Cancelled:
    penalty: int


def cancelled_from_dict(data: Dict) -> Cancelled:
    return Cancelled(penalty=data['cancel.penalty'])


@dataclass
class Job:
    config: Optional[ets2.model.JobConfig]
    started: Optional[datetime.datetime]
    ended: Optional[datetime.datetime]
    delivered: Optional[Delivered]
    cancelled: Optional[Cancelled]
    track: ets2.model.Tracks


class WorkLog:
    """A log work jobs in ETS2/ATS"""

    def __init__(self, data: ets2.model.Model):
        self._model = data
        self._model.register_observer(self)
        self._jobs: List[Job] = []

    def __repr__(self):
        return str(self._jobs)

    def notify(self, model: ets2.model.Model, change: str):
        if self._jobs:
            if model.job != self._jobs[-1].config:
                self._jobs[-1].ended = self.time_in_game(model)
                job = self.job_from_model(model)
                self._jobs.append(job)
            else:
                if model.telematic:
                    self._jobs[-1].track.add_telematic(model.telematic)
        else:
            job = self.job_from_model(model)
            self._jobs.append(job)

    def job_from_model(self, model):
        started = self.time_in_game(model)
        job = Job(config=model.job,
                  started=started,
                  ended=None,
                  delivered=None,
                  cancelled=None,
                  track=ets2.model.Tracks())
        return job

    def time_in_game(self, model):
        if model.telematic:
            started = datetime.datetime.fromordinal(1) \
                      + datetime.timedelta(minutes=model.telematic.common.game_time)
        else:
            started = None
        return started

    def job_delivered(self, delivered: Delivered) -> None:
        if self._jobs[-1].config:
            self._jobs[-1].delivered = delivered

    def job_cancelled(self, cancelled: Cancelled) -> None:
        if self._jobs[-1].config:
            self._jobs[-1].cancelled = cancelled


def add_json_to_work_log(work_log: WorkLog, json_data: json, topic: str) -> None:
    if topic == "ets2/info/gameplay/job.cancelled":
        work_log.job_cancelled(cancelled_from_dict(json_data))
    elif topic == "ets2/info/gameplay/job.delivered":
        work_log.job_delivered(delivered_from_dict(json_data))
