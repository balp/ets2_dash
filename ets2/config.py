#  Copyright (c) 2020. Anders Arnholm <Anders@Arnholm.se>
#
#   Permission to use, copy, modify, and distribute this software for any
#   purpose with or without fee is hereby granted, provided that the above
#   copyright notice and this permission notice appear in all copies.
#
#   THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
#   WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
#   MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
#   ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
#   WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
#   ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
#   OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
from dataclasses import dataclass
from typing import Dict, Optional

from ets2.types import Vector, vector_from_dict


@dataclass
class TruckConfig:
    id: str
    name: str
    adblue_capacity: float
    adblue_warning_factor: float
    battery_voltage_warning: float
    brake_air_pressure_emergency: float
    brake_air_pressure_warning: float
    brand: str
    brand_id: str
    cabin_position: Vector
    differential_ratio: float
    forward_ratio: float
    fuel_capacity: float
    fuel_warning_factor: float
    gears_forward: int
    gears_reverse: int
    head_position: Vector
    hook_position: Vector
    license_plate: str
    license_plate_country: str
    license_plate_country_id: str
    oil_pressure_warning: float
    retarder_steps: int
    reverse_ratio: float
    rpm_limit: float
    water_temperature_warning: float
    wheel_position: Vector
    wheel_powered: bool
    wheel_radius: float
    wheel_simulated: bool
    wheel_steerable: bool
    wheels_count: bool


def truckconfig_from_dict(data: Dict) -> Optional[TruckConfig]:
    if len(data):
        return TruckConfig(
            id=data['id'],
            name=data['name'],
            adblue_capacity=data['adblue.capacity'],
            adblue_warning_factor=data['adblue.warning.factor'],
            battery_voltage_warning=data['battery.voltage.warning'],
            brake_air_pressure_emergency=data['brake.air.pressure.emergency'],
            brake_air_pressure_warning=data['brake.air.pressure.warning'],
            brand=data['brand'],
            brand_id=data['brand_id'],
            cabin_position=vector_from_dict(data['cabin.position']),
            differential_ratio=data['differential.ratio'],
            forward_ratio=data['forward.ratio'],
            fuel_capacity=data['fuel.capacity'],
            fuel_warning_factor=data['fuel.warning.factor'],
            gears_forward=data['gears.forward'],
            gears_reverse=data['gears.reverse'],
            head_position=vector_from_dict(data['head.position']),
            hook_position=vector_from_dict(data['hook.position']),
            license_plate=data['license.plate'],
            license_plate_country=data['license.plate.country'],
            license_plate_country_id=data['license.plate.country.id'],
            oil_pressure_warning=data['oil.pressure.warning'],
            retarder_steps=data['retarder.steps'],
            reverse_ratio=data['reverse.ratio'],
            rpm_limit=data['rpm.limit'],
            water_temperature_warning=data['water.temperature.warning'],
            wheel_position=vector_from_dict(data['wheel.position']),
            wheel_powered=data['wheel.powered'],
            wheel_radius=data['wheel.radius'],
            wheel_simulated=data['wheel.simulated'],
            wheel_steerable=data['wheel.steerable'],
            wheels_count=data['wheels.count'],
        )
    else:
        return None


@dataclass
class TrailerConfig:
    id: str
    body_type: Optional[str]
    chain_type: Optional[str]
    cargo_accessory_id: str
    hook_position: Vector
    license_plate: str
    license_plate_country: str
    license_plate_country_id: str
    wheel_position: Vector
    wheel_powered: bool
    wheel_radius: float
    wheel_simulated: bool
    wheel_steerable: bool
    wheels_count: int


def trailer_config_from_dict(data: Dict) -> Optional[TrailerConfig]:
    if len(data):
        return TrailerConfig(
            id=data['id'],
            body_type=(data['body.type'] if 'body.type' in data else None),
            chain_type=(data['chain.type'] if 'chain.type' in data else None),
            cargo_accessory_id=data['cargo.accessory.id'],
            hook_position=vector_from_dict(data['hook.position']),
            license_plate=data['license.plate'],
            license_plate_country=data['license.plate.country'],
            license_plate_country_id=data['license.plate.country.id'],
            wheel_position=vector_from_dict(data['wheel.position']),
            wheel_powered=data['wheel.powered'],
            wheel_radius=data['wheel.radius'],
            wheel_simulated=data['wheel.simulated'],
            wheel_steerable=data['wheel.steerable'],
            wheels_count=data['wheels.count'],
        )
    else:
        return None