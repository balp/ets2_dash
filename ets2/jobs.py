from dataclasses import dataclass
from typing import Optional, Dict

import ets2.model


@dataclass
class Delivered:
    auto_load: bool
    auto_park: bool
    cargo_damage: float
    time: int
    distance: float
    xp: int
    revenue: int


def delivered_from_dict(data: Dict) -> Delivered:
    return Delivered(auto_load=data['auto.load.used'], auto_park=data['auto.park.used'],
                     cargo_damage=data['cargo.damage'], time=data['delivery.time'],
                     distance=data['distance.km'], xp=data['earned.xp'], revenue=data['revenue'])


@dataclass
class Cancelled:
    penalty: int


def cancelled_from_dict(data: Dict) -> Cancelled:
    return Cancelled(penalty=data['cancel.penalty'])


@dataclass
class Job:
    config: Optional[ets2.model.JobConfig]
    started: Optional[int]
    ended: Optional[int]
    delivered: Optional[Delivered]
    cancelled: Optional[Cancelled]
    track: ets2.model.Tracks
    id: Optional[int]
