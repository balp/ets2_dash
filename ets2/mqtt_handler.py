import json
import typing
from dataclasses import dataclass

from paho.mqtt import client as mqtt

from ets2.model import Model, add_json_to_model
from ets2_worklog.model import WorkLog, add_json_to_work_log


@dataclass
class GlobalState:
    active: bool = True


def on_message(client: typing.Any, userdata: typing.Any, message: typing.Any):
    """Handle a mqtt message and send to Model"""
    model, work_log = userdata
    assert isinstance(model, Model)
    json_data = json.loads(message.payload)
    add_json_to_model(model=model, json_data=json_data, topic=message.topic)
    if work_log is not None:
        assert isinstance(work_log, WorkLog)
        add_json_to_work_log(work_log=work_log, json_data=json_data,
                             topic=message.topic)


def mqtt_thread_loop(model: Model, work_log: WorkLog, state: GlobalState) -> None:
    client = mqtt.Client("ets2_gui", userdata=(model, work_log))
    client.on_message = on_message
    client.connect("localhost")
    client.subscribe("ets2/data")
    client.subscribe("ets2/game")
    client.subscribe("ets2/info")
    client.subscribe("ets2/info/config/job")
    client.subscribe("ets2/info/config/substances")
    # client.subscribe("ets2/info/config/controls")
    # client.subscribe("ets2/info/config/hshifter")
    client.subscribe("ets2/info/config/truck")
    # client.subscribe("ets2/info/config/trailer")

    client.subscribe("ets2/info/config/trailer.0")
    client.subscribe("ets2/info/config/trailer.1")
    client.subscribe("ets2/info/config/trailer.2")
    client.subscribe("ets2/info/config/trailer.3")
    client.subscribe("ets2/info/config/trailer.4")
    client.subscribe("ets2/info/config/trailer.5")
    client.subscribe("ets2/info/config/trailer.6")
    client.subscribe("ets2/info/config/trailer.7")
    client.subscribe("ets2/info/config/trailer.8")
    client.subscribe("ets2/info/config/trailer.9")

    client.subscribe("ets2/info/gameplay/job.cancelled")
    client.subscribe("ets2/info/gameplay/job.delivered")
    client.subscribe("ets2/info/gameplay/player.tollgate.paid")
    client.subscribe("ets2/info/gameplay/player.use.ferry")

    while state.active:
        client.loop(timeout=1.0)

