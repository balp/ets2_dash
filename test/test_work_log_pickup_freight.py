import bz2
import json

import jsonpickle as jsonpickle
from approvaltests import verify_with_namer, get_default_reporter
from approvaltests.core.namer import Namer

from ets2.model import Model, add_json_to_model
from ets2_worklog.model import WorkLog, add_json_to_work_log


def test_pickup_freight():
    work_log = rerun_data_from_files(["data/ats_start_port_ang_coos_bay.mqtt.json.bz2"])

    verify_work_log_as_json(work_log)


def test_deliver_freight():
    work_log = rerun_data_from_files(["data/ats_start_port_ang_coos_bay.mqtt.json.bz2",
                                      "data/ats_end_port_ang_coos_bay.mqtt.json.bz2"])
    verify_work_log_as_json(work_log)


def test_2_freights():
    work_log = rerun_data_from_files(["data/ats_start_port_ang_coos_bay.mqtt.json.bz2",
                                      "data/ats_end_port_ang_coos_bay.mqtt.json.bz2",
                                      "data/ats_start_coos_dalles_short.mqtt.json.bz2",
                                      "data/ats_end_coos_dalles_short.mqtt.json.bz2"
                                      ])
    verify_work_log_as_json(work_log)


def verify_work_log_as_json(work_log):
    jsonpickle.set_encoder_options('json', sort_keys=True, indent=4)
    jsonpickle.set_preferred_backend('json')
    verify_with_namer(data=jsonpickle.encode(work_log),
                      namer=Namer(extension='.json'),
                      reporter=get_default_reporter())


def rerun_data_from_files(files):
    model = Model()
    work_log: WorkLog = WorkLog(model)
    for f in files:
        with bz2.open(f) as gj:
            for line in gj:
                topic, json_str = line.decode('utf8').split(" ", 1)
                json_data = json.loads(json_str)
                add_json_to_model(model, json_data, topic)
                add_json_to_work_log(work_log, json_data, topic)
    return work_log
