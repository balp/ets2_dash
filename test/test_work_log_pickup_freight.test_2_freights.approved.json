{
    "_jobs": [
        {
            "cancelled": null,
            "config": null,
            "delivered": null,
            "ended": {
                "__reduce__": [
                    {
                        "py/type": "datetime.datetime"
                    },
                    [
                        "AAEBGhQcAAAAAA=="
                    ]
                ],
                "py/object": "datetime.datetime"
            },
            "py/object": "ets2_worklog.model.Job",
            "started": null,
            "track": {
                "_last_time": 37228,
                "points": [
                    {
                        "orientation": {
                            "heading": 0.4340660274028778,
                            "pitch": 0.0,
                            "py/object": "ets2.model.Euler",
                            "roll": 0.0
                        },
                        "position": {
                            "py/object": "ets2.model.Vector",
                            "x": -104305.18292236328,
                            "y": 12.102846145629883,
                            "z": -66655.20185089111
                        },
                        "py/object": "ets2.model.Placement"
                    },
                    {
                        "orientation": {
                            "heading": 0.9165719151496887,
                            "pitch": -4.098113549844129e-06,
                            "py/object": "ets2.model.Euler",
                            "roll": 2.504402573322295e-06
                        },
                        "position": {
                            "py/object": "ets2.model.Vector",
                            "x": -104272.02157592773,
                            "y": 12.105484962463379,
                            "z": -66604.86669540405
                        },
                        "py/object": "ets2.model.Placement"
                    }
                ],
                "py/object": "ets2.model.Tracks"
            }
        },
        {
            "cancelled": null,
            "config": {
                "cargo": "Lumber",
                "cargo_id": "lumber_b",
                "cargo_mass": 21387.240234375,
                "delivery_time": 38821,
                "destination_city": "Coos Bay",
                "destination_city_id": "coos_bay",
                "destination_company": "Olthon Homes",
                "destination_company_id": "oh_con_hom",
                "income": 18742,
                "py/object": "ets2.model.JobConfig",
                "source_city": "Port Angeles",
                "source_city_id": "port_angeles",
                "source_company": "Deepgrove",
                "source_company_id": "dg_wd_saw"
            },
            "delivered": {
                "auto_load": true,
                "auto_park": false,
                "cargo_damage": 0.0,
                "distance": 676.0,
                "py/object": "ets2_worklog.model.Delivered",
                "revenue": 18742,
                "time": 1084,
                "xp": 870
            },
            "ended": {
                "__reduce__": [
                    {
                        "py/type": "datetime.datetime"
                    },
                    [
                        "AAEBGw4gAAAAAA=="
                    ]
                ],
                "py/object": "datetime.datetime"
            },
            "py/object": "ets2_worklog.model.Job",
            "started": {
                "__reduce__": [
                    {
                        "py/type": "datetime.datetime"
                    },
                    [
                        "AAEBGhQcAAAAAA=="
                    ]
                ],
                "py/object": "datetime.datetime"
            },
            "track": {
                "_last_time": 38312,
                "points": [
                    {
                        "py/id": 9
                    },
                    {
                        "orientation": {
                            "heading": 0.38147035241127014,
                            "pitch": 0.00024254759773612022,
                            "py/object": "ets2.model.Euler",
                            "roll": 6.837778346380219e-05
                        },
                        "position": {
                            "py/object": "ets2.model.Vector",
                            "x": -113779.6833114624,
                            "y": 5.177712440490723,
                            "z": -42559.202308654785
                        },
                        "py/object": "ets2.model.Placement"
                    },
                    {
                        "orientation": {
                            "heading": 0.3937177062034607,
                            "pitch": 8.139308192767203e-05,
                            "py/object": "ets2.model.Euler",
                            "roll": -1.51781978274812e-06
                        },
                        "position": {
                            "py/object": "ets2.model.Vector",
                            "x": -113934.7191772461,
                            "y": 5.177143096923828,
                            "z": -42777.36331176758
                        },
                        "py/object": "ets2.model.Placement"
                    }
                ],
                "py/object": "ets2.model.Tracks"
            }
        },
        {
            "cancelled": null,
            "config": null,
            "delivered": null,
            "ended": {
                "__reduce__": [
                    {
                        "py/type": "datetime.datetime"
                    },
                    [
                        "AAEBGw4kAAAAAA=="
                    ]
                ],
                "py/object": "datetime.datetime"
            },
            "py/object": "ets2_worklog.model.Job",
            "started": {
                "__reduce__": [
                    {
                        "py/type": "datetime.datetime"
                    },
                    [
                        "AAEBGw4gAAAAAA=="
                    ]
                ],
                "py/object": "datetime.datetime"
            },
            "track": {
                "_last_time": 38316,
                "points": [
                    {
                        "orientation": {
                            "heading": 0.3937177062034607,
                            "pitch": 8.226583304349333e-05,
                            "py/object": "ets2.model.Euler",
                            "roll": -1.51781978274812e-06
                        },
                        "position": {
                            "py/object": "ets2.model.Vector",
                            "x": -113934.7191772461,
                            "y": 5.177135944366455,
                            "z": -42777.36331176758
                        },
                        "py/object": "ets2.model.Placement"
                    },
                    {
                        "orientation": {
                            "heading": 0.6608810424804688,
                            "pitch": -6.7163523453928065e-06,
                            "py/object": "ets2.model.Euler",
                            "roll": 5.54004191144486e-06
                        },
                        "position": {
                            "py/object": "ets2.model.Vector",
                            "x": -113746.27894592285,
                            "y": 5.177285194396973,
                            "z": -42803.44400024414
                        },
                        "py/object": "ets2.model.Placement"
                    },
                    {
                        "orientation": {
                            "heading": 0.42154666781425476,
                            "pitch": -0.0001929907884914428,
                            "py/object": "ets2.model.Euler",
                            "roll": -0.00018251781875733286
                        },
                        "position": {
                            "py/object": "ets2.model.Vector",
                            "x": -113731.70094299316,
                            "y": 5.183038234710693,
                            "z": -42779.22634887695
                        },
                        "py/object": "ets2.model.Placement"
                    }
                ],
                "py/object": "ets2.model.Tracks"
            }
        },
        {
            "cancelled": null,
            "config": {
                "cargo": "Gypsum",
                "cargo_id": "gypsum",
                "cargo_mass": 21492.798828125,
                "delivery_time": 38977,
                "destination_city": "The Dalles",
                "destination_city_id": "the_dalles",
                "destination_company": "Home Store",
                "destination_company_id": "hs_mkt",
                "income": 12429,
                "py/object": "ets2.model.JobConfig",
                "source_city": "Coos Bay",
                "source_city_id": "coos_bay",
                "source_company": "Home Store",
                "source_company_id": "hs_whs"
            },
            "delivered": {
                "auto_load": true,
                "auto_park": false,
                "cargo_damage": 0.0,
                "distance": 487.0,
                "py/object": "ets2_worklog.model.Delivered",
                "revenue": 12429,
                "time": 351,
                "xp": 795
            },
            "ended": {
                "__reduce__": [
                    {
                        "py/type": "datetime.datetime"
                    },
                    [
                        "AAEBGxQbAAAAAA=="
                    ]
                ],
                "py/object": "datetime.datetime"
            },
            "py/object": "ets2_worklog.model.Job",
            "started": {
                "__reduce__": [
                    {
                        "py/type": "datetime.datetime"
                    },
                    [
                        "AAEBGw4kAAAAAA=="
                    ]
                ],
                "py/object": "datetime.datetime"
            },
            "track": {
                "_last_time": 38667,
                "points": [
                    {
                        "py/id": 36
                    },
                    {
                        "orientation": {
                            "heading": 0.90943443775177,
                            "pitch": 1.753081778588239e-05,
                            "py/object": "ets2.model.Euler",
                            "roll": -1.13836483706109e-06
                        },
                        "position": {
                            "py/object": "ets2.model.Vector",
                            "x": -113712.37997817993,
                            "y": 5.176380634307861,
                            "z": -42736.49880981445
                        },
                        "py/object": "ets2.model.Placement"
                    },
                    {
                        "orientation": {
                            "heading": 0.3180747926235199,
                            "pitch": -0.0003237130003981292,
                            "py/object": "ets2.model.Euler",
                            "roll": -0.0015823650173842907
                        },
                        "position": {
                            "py/object": "ets2.model.Vector",
                            "x": -99166.45639038086,
                            "y": 12.230788230895996,
                            "z": -51001.42868041992
                        },
                        "py/object": "ets2.model.Placement"
                    },
                    {
                        "orientation": {
                            "heading": 0.2571701109409332,
                            "pitch": 0.0020690918900072575,
                            "py/object": "ets2.model.Euler",
                            "roll": 0.0001236264215549454
                        },
                        "position": {
                            "py/object": "ets2.model.Vector",
                            "x": -99233.9595336914,
                            "y": 12.250280380249023,
                            "z": -50934.83279418945
                        },
                        "py/object": "ets2.model.Placement"
                    },
                    {
                        "orientation": {
                            "heading": 0.2233603447675705,
                            "pitch": 0.00035232389927841723,
                            "py/object": "ets2.model.Euler",
                            "roll": -5.710796904168092e-05
                        },
                        "position": {
                            "py/object": "ets2.model.Vector",
                            "x": -99288.8861694336,
                            "y": 12.376511573791504,
                            "z": -51052.101501464844
                        },
                        "py/object": "ets2.model.Placement"
                    }
                ],
                "py/object": "ets2.model.Tracks"
            }
        },
        {
            "cancelled": null,
            "config": null,
            "delivered": null,
            "ended": null,
            "py/object": "ets2_worklog.model.Job",
            "started": {
                "__reduce__": [
                    {
                        "py/type": "datetime.datetime"
                    },
                    [
                        "AAEBGxQbAAAAAA=="
                    ]
                ],
                "py/object": "datetime.datetime"
            },
            "track": {
                "_last_time": 38667,
                "points": [
                    {
                        "orientation": {
                            "heading": 0.22336062788963318,
                            "pitch": 0.0003549421380739659,
                            "py/object": "ets2.model.Euler",
                            "roll": -5.7145913160638884e-05
                        },
                        "position": {
                            "py/object": "ets2.model.Vector",
                            "x": -99288.8861694336,
                            "y": 12.376470565795898,
                            "z": -51052.101501464844
                        },
                        "py/object": "ets2.model.Placement"
                    }
                ],
                "py/object": "ets2.model.Tracks"
            }
        }
    ],
    "_model": {
        "_observers": [
            {
                "py/id": 0
            }
        ],
        "game": {
            "id": "ats",
            "name": "American Truck Simulator 1.36.1.34s",
            "py/object": "ets2.model.Game",
            "raw_version": 65539,
            "version": {
                "major": 1,
                "minor": 3,
                "py/object": "ets2.model.Version"
            }
        },
        "info": {
            "paused": true,
            "py/object": "ets2.model.Info"
        },
        "job": null,
        "py/object": "ets2.model.Model",
        "telematic": {
            "common": {
                "game_time": 38667,
                "py/object": "ets2.model.Common",
                "rest_stop": 226,
                "scale": 3.0
            },
            "py/object": "ets2.model.Telematic",
            "trailer": {
                "acceleration_angular": {
                    "py/object": "ets2.model.Vector",
                    "x": -0.0005581545410677791,
                    "y": -1.2042250091326423e-05,
                    "z": -3.1988733098842204e-05
                },
                "acceleration_linear": {
                    "py/object": "ets2.model.Vector",
                    "x": 0.00046772247878834605,
                    "y": -0.013350364752113819,
                    "z": 3.2667260256857844e-06
                },
                "cargo_damage": 2.6963388677359035e-07,
                "connected": false,
                "py/object": "ets2.model.Trailer",
                "velocity_angular": {
                    "py/object": "ets2.model.Vector",
                    "x": 6.509331433335319e-05,
                    "y": -7.756915692880284e-06,
                    "z": -7.646862059118575e-07
                },
                "velocity_linear": {
                    "py/object": "ets2.model.Vector",
                    "x": 0.0002764087694231421,
                    "y": 0.002197939669713378,
                    "z": -3.936955181416124e-05
                },
                "wear_chassis": 0.009722059592604637,
                "wear_wheels": 0.009722059592604637,
                "world_placement": {
                    "orientation": {
                        "heading": 0.24393245577812195,
                        "pitch": -5.5210693972185254e-05,
                        "py/object": "ets2.model.Euler",
                        "roll": -2.117358599207364e-05
                    },
                    "position": {
                        "py/object": "ets2.model.Vector",
                        "x": -99282.47064208984,
                        "y": 12.375089645385742,
                        "z": -51051.71896362305
                    },
                    "py/object": "ets2.model.Placement"
                }
            },
            "truck": {
                "adblue": 54.08164978027344,
                "adblue_warning": false,
                "battery_voltage": 13.688865661621094,
                "battery_voltage_warning": false,
                "brake_air_pressure": 88.40076446533203,
                "brake_air_pressure_emergency": false,
                "brake_air_pressure_warning": false,
                "brake_motor": false,
                "brake_parking": false,
                "brake_retarder": 0,
                "brake_temperature": 61.57093048095703,
                "cabin_acceleration_angular": {
                    "py/object": "ets2.model.Vector",
                    "x": 0.0,
                    "y": 0.0,
                    "z": 0.0
                },
                "cabin_offset": {
                    "orientation": {
                        "heading": 0.0,
                        "pitch": 0.0,
                        "py/object": "ets2.model.Euler",
                        "roll": 0.0
                    },
                    "position": {
                        "py/object": "ets2.model.Vector",
                        "x": 0.0,
                        "y": 0.0,
                        "z": 0.0
                    },
                    "py/object": "ets2.model.Placement"
                },
                "cabin_velocity_angular": {
                    "py/object": "ets2.model.Vector",
                    "x": 0.0,
                    "y": 0.0,
                    "z": 0.0
                },
                "cruise_control": 0.0,
                "dashboard_backlight": 1.0,
                "displayed_gear": -1,
                "effective_brake": 1.0,
                "effective_clutch": 0.0,
                "effective_steering": 0.567232608795166,
                "effective_throttle": 0.0,
                "electric_enabled": true,
                "engine_enabled": true,
                "engine_gear": 0,
                "engine_rpm": 599.9972534179688,
                "fuel_amount": 239.63307189941406,
                "fuel_consumption_average": 0.46449264883995056,
                "fuel_range": 752.4710083007812,
                "fuel_warning": false,
                "head_offset": {
                    "orientation": {
                        "heading": 0.0,
                        "pitch": -0.01944444701075554,
                        "py/object": "ets2.model.Euler",
                        "roll": 0.0
                    },
                    "position": {
                        "py/object": "ets2.model.Vector",
                        "x": 0.0,
                        "y": -0.019999999552965164,
                        "z": 0.0
                    },
                    "py/object": "ets2.model.Placement"
                },
                "hshifter_slot": 0,
                "input_brake": 0.0,
                "input_clutch": 0.0,
                "input_steering": 0.567232608795166,
                "input_throttle": 0.0,
                "lblinker": false,
                "light_aux_front": 1,
                "light_aux_roof": 1,
                "light_beacon": false,
                "light_beam_high": false,
                "light_beam_low": true,
                "light_brake": false,
                "light_lblinker": false,
                "light_parking": true,
                "light_rblinker": false,
                "light_reverse": true,
                "local_acceleration_angular": {
                    "py/object": "ets2.model.Vector",
                    "x": -0.001948226010426879,
                    "y": -2.1071453375043347e-05,
                    "z": 9.130919352173805e-05
                },
                "local_acceleration_linear": {
                    "py/object": "ets2.model.Vector",
                    "x": 0.0002647233195602894,
                    "y": 0.03982933610677719,
                    "z": 0.00021258782362565398
                },
                "local_velocity_angular": {
                    "py/object": "ets2.model.Vector",
                    "x": 0.0001670177443884313,
                    "y": 1.512439575890312e-05,
                    "z": -1.0040561164714745e-06
                },
                "local_velocity_linear": {
                    "py/object": "ets2.model.Vector",
                    "x": -1.6030235201469623e-05,
                    "y": -0.0015500109875574708,
                    "z": 0.000493462139274925
                },
                "navigation_distance": 393.462646484375,
                "navigation_speed_limit": 0.0,
                "navigation_time": 22.875734329223633,
                "odometer": 17625.7421875,
                "oil_pressure": 39.545352935791016,
                "oil_pressure_warning": false,
                "oil_temperature": 91.64779663085938,
                "py/object": "ets2.model.Truck",
                "rblinker": false,
                "speed": -0.01175117027014494,
                "water_temperature": 87.06563568115234,
                "water_temperature_warning": false,
                "wear_cabin": 0.0025513770524412394,
                "wear_chassis": 0.0031892212573438883,
                "wear_engine": 0.004692628979682922,
                "wear_transmission": 0.0040547847747802734,
                "wear_wheels": 0.03145786374807358,
                "wipers": false,
                "world_placement": {
                    "py/id": 62
                }
            }
        },
        "tracks": {
            "_last_time": 38667,
            "points": [
                {
                    "py/id": 6
                },
                {
                    "py/id": 9
                },
                {
                    "py/id": 19
                },
                {
                    "py/id": 22
                },
                {
                    "py/id": 33
                },
                {
                    "py/id": 36
                },
                {
                    "py/id": 46
                },
                {
                    "py/id": 49
                },
                {
                    "py/id": 52
                },
                {
                    "py/id": 55
                }
            ],
            "py/object": "ets2.model.Tracks"
        },
        "trailer_config": [
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null
        ],
        "truck_config": {
            "adblue_capacity": 80.0,
            "adblue_warning_factor": 0.15000000596046448,
            "battery_voltage_warning": 11.880000114440918,
            "brake_air_pressure_emergency": 29.579999923706055,
            "brake_air_pressure_warning": 59.15999984741211,
            "brand": "Volvo",
            "brand_id": "volvo",
            "cabin_position": {
                "py/object": "ets2.model.Vector",
                "x": 0.0,
                "y": 3.0,
                "z": -2.0
            },
            "differential_ratio": 3.549999952316284,
            "forward_ratio": 0.7300000190734863,
            "fuel_capacity": 758.0,
            "fuel_warning_factor": 0.15000000596046448,
            "gears_forward": 13,
            "gears_reverse": 3,
            "head_position": {
                "py/object": "ets2.model.Vector",
                "x": -0.5642366409301758,
                "y": -0.5814938545227051,
                "z": 0.570602536201477
            },
            "hook_position": {
                "py/object": "ets2.model.Vector",
                "x": 0.0,
                "y": 1.0,
                "z": 1.068235158920288
            },
            "id": "vehicle.volvo.vnl",
            "license_plate": "8A7P001",
            "license_plate_country": "California",
            "license_plate_country_id": "california",
            "name": "VNL",
            "oil_pressure_warning": 10.149999618530273,
            "py/object": "ets2.model.TruckConfig",
            "retarder_steps": 3,
            "reverse_ratio": -4.03000020980835,
            "rpm_limit": 2100.0,
            "water_temperature_warning": 105.0,
            "wheel_position": {
                "py/object": "ets2.model.Vector",
                "x": 0.9612910747528076,
                "y": 0.5059999823570251,
                "z": 0.5031635761260986
            },
            "wheel_powered": true,
            "wheel_radius": 0.5072842240333557,
            "wheel_simulated": true,
            "wheel_steerable": false,
            "wheels_count": 6
        }
    },
    "py/object": "ets2_worklog.model.WorkLog"
}